File: NC/src/RichardsonExtrapolation.cpp
#include "../include/RichardsonExtrapolation.hpp"
#include "../include/CentralDifference.hpp"

RichardsonExtrapolation::RichardsonExtrapolation(function<double(double)> f, double h)
    : Differentiation(f, h) {}

double RichardsonExtrapolation::centralAt(double x, double stepSize) const {
    CentralDifference cd(func, stepSize);
    return cd.derivative(x);
}

double RichardsonExtrapolation::derivative(double x) const {
    double D_h    = centralAt(x, h);
    double D_half = centralAt(x, h / 2.0);
    return (4.0 * D_half - D_h) / 3.0;
}

File: Numerical_Computing_2/src/differentiation/richardson_extrapolation.py
from .central_difference import CentralDifference
from .numerical_differentiation import NumericalDifferentiation


class RichardsonExtrapolation(NumericalDifferentiation):
    def central_difference(self, x: float, h: float) -> float:
        return CentralDifference(self.func, h).derivative(x)

    def derivative(self, x: float) -> float:
        d_h = self.central_difference(x, self.h)
        d_half = self.central_difference(x, self.h / 2.0)
        return (4.0 * d_half - d_h) / 3.0

    @property
    def method_name(self) -> str:
        return "Richardson Extrapolation"
